aws/alb.go

```diff
@@ -0,0 +1,219 @@
+	// TODO: paged results
+	resp, err := elbv2.DescribeLoadBalancers(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: filter for ALBs with a given set of tags? For ex.: KubernetesCluster=foo
+	for _, lb := range resp.LoadBalancers {
+		listeners, err := getListeners(elbv2, aws.StringValue(lb.LoadBalancerArn))
+		if err != nil {
+			log.Printf("failed to describe listeners for load balancer %q: %v\n", lb.LoadBalancerName, err)
+			continue
+		}
+		for _, listener := range listeners {
+			for _, cert := range listener.Certificates {
+				certARN := aws.StringValue(cert.CertificateArn)
+				if certARN == certificateARN {
+					return &LoadBalancer{
+						name: aws.StringValue(lb.LoadBalancerName),
+						arn:  aws.StringValue(lb.LoadBalancerArn),
+						listener: &loadBalancerListener{
+							port:           aws.Int64Value(listener.Port),
+							arn:            aws.StringValue(listener.ListenerArn),
+							certificateARN: certARN,
+						},
+					}, nil
+				}
+			}
+		}
+	}
+
+	return nil, ErrLoadBalancerNotFound
+}
+
+func getListeners(alb elbv2iface.ELBV2API, loadBalancerARN string) ([]*elbv2.Listener, error) {
+	// TODO: paged results
+	params := &elbv2.DescribeListenersInput{
+		LoadBalancerArn: aws.String(loadBalancerARN),
+	}
+	resp, err := alb.DescribeListeners(params)
+
+	if err != nil {
+		return nil, err
+	}
+	return resp.Listeners, nil
+}
+
+type createLoadBalancerSpec struct {
+	name            string
+	scheme          string
+	subnets         []string
+	certificateARN  string
+	securityGroupID string
+	clusterName     string
+	targetGroupARNs []string
+}
+
+func createLoadBalancer(alb elbv2iface.ELBV2API, spec *createLoadBalancerSpec) (*LoadBalancer, error) {
+	var name = normalizeLoadBalancerName(spec.certificateARN)
+	params := &elbv2.CreateLoadBalancerInput{
+		Name:    aws.String(name),
+		Subnets: aws.StringSlice(spec.subnets),
+		Scheme:  aws.String(spec.scheme),
+		SecurityGroups: []*string{
+			aws.String(spec.securityGroupID),
+		},
+		Tags: []*elbv2.Tag{
+			{
+				Key:   aws.String(kubernetesClusterTag),
+				Value: aws.String(spec.clusterName),
+			},
+			{
+				Key:   aws.String(kubernetesCreatorTag),
+				Value: aws.String("kube-ingress-aws-controller"),
+			},
+		},
+	}
+	resp, err := alb.CreateLoadBalancer(params)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.LoadBalancers) < 1 {
+		return nil, errors.New("request to create ALB succeeded but returned no items")
+	}
+
+	newLoadBalancer := resp.LoadBalancers[0]
+	loadBalancerARN := aws.StringValue(newLoadBalancer.LoadBalancerArn)
+	newListener, err := createListener(alb, loadBalancerARN, spec)
+	if err != nil {
+		// TODO: delete just created LB?
+		return nil, err
+	}
+
+	return &LoadBalancer{
+		arn:      loadBalancerARN,
+		name:     name,
+		dnsName:  aws.StringValue(newLoadBalancer.DNSName),
+		listener: newListener,
+	}, nil
+}
+
+func normalizeLoadBalancerName(name string) string {
+	fields := strings.Split(name, "/")
```

Comment:
if it is limited what characters can appear in the name, then i would use a regular expression replace, to make sure that only the allowed characters appear in the name. As a replacing character, i would use something like '__' (double underscore), because simply removing characters may result in conflicts.